Statically compiling `ctr` fails on Go 1.15 and up (segfault or not statically linked)

It seems no longer possible to produce a static binary for `ctr`.

Relates to:

- https://github.com/golang/go/issues/40711 cmd/go: can't build a static binary since Go 1.15
- https://github.com/containerd/containerd/pull/4470 BUILDING.md: fix description about static builds


Using the following Dockerfile:

```dockerfile
ARG GO_VERSION=1.16.6
FROM golang:${GO_VERSION}-buster AS build
ENV GO111MODULE=off
RUN apt-get update && apt-get install -y --no-install-recommends libbtrfs-dev
RUN git clone https://github.com/containerd/containerd.git "/go/src/github.com/containerd/containerd"
WORKDIR /go/src/github.com/containerd/containerd
RUN git checkout -q "v1.5.5"
RUN make EXTRA_FLAGS="-buildmode pie" \
  EXTRA_LDFLAGS='-linkmode external -extldflags "-fno-PIC -static"' \
  BUILDTAGS="netgo osusergo static_build" \
  bin/ctr
RUN bin/ctr --version
RUN ldd bin/ctr || echo "statically linked?"

FROM alpine:3.13
COPY --from=build /go/src/github.com/containerd/containerd/bin/. /usr/local/bin/
RUN ldd /usr/local/bin/ctr || echo "statically linked?"
RUN /usr/local/bin/ctr --version
```

Building with Go 1.14.15 succeeds, and produces a static binary:

```bash
docker build -t foo --build-arg GO_VERSION=1.14.15 .
...
Step 9/14 : RUN bin/ctr --version
 ---> Running in 23134fe72c41
ctr github.com/containerd/containerd v1.5.5
Removing intermediate container 23134fe72c41
 ---> de6f6931da51
Step 10/14 : RUN ldd bin/ctr || echo "statically linked"
 ---> Running in b9c480ab893c
	not a dynamic executable
statically linked?
Removing intermediate container b9c480ab893c
 ---> 2fc3c219f7b1
Step 11/14 : FROM alpine:3.13
 ---> 6dbb9cc54074
Step 12/14 : COPY --from=build /go/src/github.com/containerd/containerd/bin/. /usr/local/bin/
 ---> d6b7982e3e7f
Step 13/14 : RUN ldd /usr/local/bin/ctr || echo "statically linked"
 ---> Running in 9a24797aae98
/lib/ld-musl-x86_64.so.1: /usr/local/bin/ctr: Not a valid dynamic program
statically linked?
Removing intermediate container 9a24797aae98
 ---> 63b53120e884
Step 14/14 : RUN /usr/local/bin/ctr --version
 ---> Running in d5b658b4a53d
ctr github.com/containerd/containerd v1.5.5
Removing intermediate container d5b658b4a53d
 ---> 3153084ff76c
Successfully built 3153084ff76c
Successfully tagged foo:latest
```

Building with Go 1.15 or Go 1.16 fails; the `ctr` binary segfaults:

```
docker build -t foo --build-arg GO_VERSION=1.16.6 .
Step 8/14 : RUN make EXTRA_FLAGS="-buildmode pie"   EXTRA_LDFLAGS='-linkmode external -extldflags "-fno-PIC -static"'   BUILDTAGS="netgo osusergo static_build"   bin/ctr
 ---> Running in cc5830fe6e3d
+ bin/ctr
# github.com/containerd/containerd/cmd/ctr
loadinternal: cannot find runtime/cgo
Removing intermediate container cc5830fe6e3d
 ---> 0b9462d31283
Step 9/14 : RUN bin/ctr --version
 ---> Running in 70a3c28fb31f
Trace/breakpoint trap (core dumped)
The command '/bin/sh -c bin/ctr --version' returned a non-zero code: 133
```

The binary produced is defunct, and outputs `Trace/breakpoint trap`, similar to what's reported long time ago in https://github.com/golang/go/issues/14795 (related: https://github.com/moby/moby/pull/22580)

Removing `-buildmode pie` results in the same:

```
Step 8/14 : RUN make EXTRA_FLAGS=" "   EXTRA_LDFLAGS='-linkmode external -extldflags "-fno-PIC -static"'   BUILDTAGS="netgo osusergo static_build"   bin/ctr
 ---> Running in 310fe840f6c9
+ bin/ctr
# github.com/containerd/containerd/cmd/ctr
loadinternal: cannot find runtime/cgo
Removing intermediate container 310fe840f6c9
 ---> 39c65f6d5847
Step 9/14 : RUN bin/ctr --version
 ---> Running in 54ca078b2ebb
Trace/breakpoint trap (core dumped)
The command '/bin/sh -c bin/ctr --version' returned a non-zero code: 133
```

Removing `-linkmode external` makes the `build` stage work, but does not produce a statically linked binary `ldd` is incorrect here; it reports the binary to be static, but it's not; copying the binary to an `alpine` image shows that it's not a statically linked binary. Removing both `--buildmode=pie` and `-linkmode external` results in the same (not statically linked).

```dockerfile
Step 8/14 : RUN make EXTRA_FLAGS="-buildmode pie"   EXTRA_LDFLAGS='-extldflags "-fno-PIC -static"'   BUILDTAGS="netgo osusergo static_build"   bin/ctr
 ---> Running in c41e2e6b1e83
+ bin/ctr
Removing intermediate container c41e2e6b1e83
 ---> 3997edbfcab9
Step 9/14 : RUN bin/ctr --version
 ---> Running in 557e71c84967
ctr github.com/containerd/containerd v1.5.5
Removing intermediate container 557e71c84967
 ---> cb17a8b95aa0
Step 10/14 : RUN ldd bin/ctr || echo "statically linked?"
 ---> Running in cbc61dcada62
	statically linked?
Removing intermediate container cbc61dcada62
 ---> fbefa0076f01
Step 11/14 : FROM alpine:3.13
 ---> 6dbb9cc54074
Step 12/14 : COPY --from=build /go/src/github.com/containerd/containerd/bin/. /usr/local/bin/
 ---> c5d6edff351c
Step 13/14 : RUN ldd /usr/local/bin/ctr || echo "statically linked?"
 ---> Running in 8c7b1be6c6b9
	/lib64/ld-linux-x86-64.so.2 (0x7efce75d9000)
Removing intermediate container 8c7b1be6c6b9
 ---> f13564163f06
Step 14/14 : RUN /usr/local/bin/ctr --version
 ---> Running in 52d9aa126d58
/bin/sh: /usr/local/bin/ctr: not found
The command '/bin/sh -c /usr/local/bin/ctr --version' returned a non-zero code: 127
```

Mixing `-fno-PIC` and `-buildmode pie` looks suspicious for me. The below flags works for me.

```
[ec2-user@ip-172-31-4-136 containerd]$ make EXTRA_FLAGS="-buildmode pie" EXTRA_L
DFLAGS="-linkmode external -extldflags -static" 'BUILDTAGS=netgo osusergo static
_build no_devmapper no_btrfs no_aufs no_zfs'
+ bin/ctr
go build  -gcflags=-trimpath=/src -buildmode=pie  -o bin/ctr -ldflags '-X github.com/containerd/containerd/version.Version=v1.5.0-347-g218db0f9a.m -X github.com/containerd/containerd/version.Revision=218db0f9afaeda2813bddbddf2764c8414bb9c42.m -X github.com/containerd/containerd/version.Package=github.com/containerd/containerd -linkmode external -extldflags -static' -tags "netgo osusergo static_build no_devmapper no_btrfs no_aufs no_zfs"  ./cmd/ctr
+ bin/containerd
go build  -gcflags=-trimpath=/src -buildmode=pie  -o bin/containerd -ldflags '-X github.com/containerd/containerd/version.Version=v1.5.0-347-g218db0f9a.m -X github.com/containerd/containerd/version.Revision=218db0f9afaeda2813bddbddf2764c8414bb9c42.m -X github.com/containerd/containerd/version.Package=github.com/containerd/containerd -linkmode external -extldflags -static' -tags "netgo osusergo static_build no_devmapper no_btrfs no_aufs no_zfs"  ./cmd/containerd
# github.com/containerd/containerd/cmd/containerd
/tmp/go-link-995192107/000002.o: In function `New':
/home/ec2-user/go/src/github.com/containerd/containerd/vendor/github.com/miekg/pkcs11/pkcs11.go:75: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
+ bin/containerd-stress
go build  -gcflags=-trimpath=/src -buildmode=pie  -o bin/containerd-stress -ldflags '-X github.com/containerd/containerd/version.Version=v1.5.0-347-g218db0f9a.m -X github.com/containerd/containerd/version.Revision=218db0f9afaeda2813bddbddf2764c8414bb9c42.m -X github.com/containerd/containerd/version.Package=github.com/containerd/containerd -linkmode external -extldflags -static' -tags "netgo osusergo static_build no_devmapper no_btrfs no_aufs no_zfs"  ./cmd/containerd-stress
+ bin/containerd-shim
# github.com/containerd/containerd/cmd/containerd-shim
loadinternal: cannot find runtime/cgo
+ bin/containerd-shim-runc-v1
# github.com/containerd/containerd/cmd/containerd-shim-runc-v1
loadinternal: cannot find runtime/cgo
+ bin/containerd-shim-runc-v2
# github.com/containerd/containerd/cmd/containerd-shim-runc-v2
loadinternal: cannot find runtime/cgo
+ binaries
[ec2-user@ip-172-31-4-136 containerd]$ 
```

Then

```
[ec2-user@ip-172-31-4-136 containerd]$ ldd bin/ctr
        not a dynamic executable
[ec2-user@ip-172-31-4-136 containerd]$ ./bin/ctr --version
ctr github.com/containerd/containerd v1.5.0-347-g218db0f9a.m
[ec2-user@ip-172-31-4-136 containerd]$ 
```

--- a/BUILDING.md
+++ b/BUILDING.md
@@ -122,14 +122,13 @@ Please refer to [RUNC.md](/docs/RUNC.md) for the currently supported version of 
 You can build static binaries by providing a few variables to `make`:
 
 ```sh
-make EXTRA_FLAGS="-buildmode pie" \
-	EXTRA_LDFLAGS='-linkmode external -extldflags "-fno-PIC -static"' \
-	BUILDTAGS="netgo osusergo static_build"
+make STATIC=1
 ```
 
 > *Note*:
 > - static build is discouraged
 > - static containerd binary does not support loading shared object plugins (`*.so`)
+> - static build binaries are not position-independent
 
 # Via Docker container
 

--- a/Makefile
+++ b/Makefile
@@ -90,8 +90,17 @@ endif
 GO_BUILDTAGS ?=
 GO_BUILDTAGS += urfave_cli_no_docs
 GO_BUILDTAGS += ${DEBUG_TAGS}
+ifneq ($(STATIC),)
+	GO_BUILDTAGS += osusergo netgo static_build
+endif
 GO_TAGS=$(if $(GO_BUILDTAGS),-tags "$(strip $(GO_BUILDTAGS))",)
-GO_LDFLAGS=-ldflags '-X $(PKG)/version.Version=$(VERSION) -X $(PKG)/version.Revision=$(REVISION) -X $(PKG)/version.Package=$(PACKAGE) $(EXTRA_LDFLAGS)'
+
+GO_LDFLAGS=-ldflags '-X $(PKG)/version.Version=$(VERSION) -X $(PKG)/version.Revision=$(REVISION) -X $(PKG)/version.Package=$(PACKAGE) $(EXTRA_LDFLAGS)
+ifneq ($(STATIC),)
+	GO_LDFLAGS += -extldflags "-static"
+endif
+GO_LDFLAGS+='
+
 SHIM_GO_LDFLAGS=-ldflags '-X $(PKG)/version.Version=$(VERSION) -X $(PKG)/version.Revision=$(REVISION) -X $(PKG)/version.Package=$(PACKAGE) -extldflags "-static" $(EXTRA_LDFLAGS)'
 
 # Project packages.

--- a/Makefile.linux
+++ b/Makefile.linux
@@ -21,7 +21,9 @@ COMMANDS += containerd-shim containerd-shim-runc-v1 containerd-shim-runc-v2
 # check GOOS for cross compile builds
 ifeq ($(GOOS),linux)
   ifneq ($(GOARCH),$(filter $(GOARCH),mips mipsle mips64 mips64le ppc64))
-	GO_GCFLAGS += -buildmode=pie
+    ifeq ($(STATIC),)
+	    GO_GCFLAGS += -buildmode=pie
+    endif
   endif
 endif
 

